File: scripts/analysis/analyze_module_deps.py
#!/usr/bin/env python3
"""
Module dependency analyzer for franken-engine

This script analyzes the dependency graph of all modules in the franken-engine crate
and classifies them based on reachability from entry points.
"""

import os
import re
import json
from collections import defaultdict, deque
from typing import Dict, Set, List, Tuple

def collect_modules(src_path: str) -> Dict[str, str]:
    """Collect all Rust modules in the source directory."""
    modules = {}

    # Collect main modules
    for filename in os.listdir(src_path):
        if filename.endswith('.rs'):
            module_name = filename[:-3]  # Remove .rs extension
            modules[module_name] = os.path.join(src_path, filename)

    # Collect binary targets
    bin_path = os.path.join(src_path, 'bin')
    if os.path.exists(bin_path):
        for filename in os.listdir(bin_path):
            if filename.endswith('.rs'):
                module_name = f"bin_{filename[:-3]}"
                modules[module_name] = os.path.join(bin_path, filename)

    return modules

def parse_module_dependencies(file_path: str) -> Set[str]:
    """Parse dependencies from a Rust module file."""
    deps = set()

    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
    except:
        return deps

    # Match use statements that reference crate modules
    # Look for patterns like: use crate::module_name or use super::module_name
    use_patterns = [
        r'use\s+crate::([a-zA-Z_][a-zA-Z0-9_]*)',  # use crate::module
        r'use\s+super::([a-zA-Z_][a-zA-Z0-9_]*)',  # use super::module
        r'use\s+([a-zA-Z_][a-zA-Z0-9_]*)::', # use module::
        r'pub\s+mod\s+([a-zA-Z_][a-zA-Z0-9_]*)', # pub mod module
        r'mod\s+([a-zA-Z_][a-zA-Z0-9_]*)', # mod module
    ]

    for pattern in use_patterns:
        matches = re.findall(pattern, content)
        for match in matches:
            if isinstance(match, tuple):
                deps.add(match[0])
            else:
                deps.add(match)

    return deps

def get_entry_points(modules: Dict[str, str]) -> Set[str]:
    """Define the entry points for dependency analysis."""
    entry_points = set()

    # Primary entry points
    entry_points.add("execution_orchestrator")
    entry_points.add("lib")

    # Binary targets are also entry points
    for module_name in modules.keys():
        if module_name.startswith("bin_"):
            entry_points.add(module_name)

    return entry_points

def compute_reachable_modules(entry_point: str, module_deps: Dict[str, Set[str]]) -> Set[str]:
    """Compute all modules reachable from a given entry point using BFS."""
    visited = set()
    to_visit = deque([entry_point])

    while to_visit:
        current = to_visit.popleft()

        if current in visited:
            continue

        visited.add(current)

        # Add dependencies to visit queue
        if current in module_deps:
            for dep in module_deps[current]:
                if dep not in visited:
                    to_visit.append(dep)

    return visited

def classify_module(module_name: str, reachable_from: Set[str]) -> str:
    """Classify a module based on which entry points can reach it."""
    if not reachable_from:
        return "island"

    # Core reachable: reachable from execution_orchestrator or lib
    if "execution_orchestrator" in reachable_from or "lib" in reachable_from:
        return "core_reachable"

    # Check if module name suggests governance/control plane
    governance_indicators = [
        "control_plane", "governance", "policy", "scorecard", "audit",
        "compliance", "security", "attestation", "witness", "evidence"
    ]

    gate_indicators = [
        "gate", "verification", "conformance", "test", "benchmark",
        "harness", "validator", "checker", "runner", "oracle"
    ]

    module_lower = module_name.lower()

    for indicator in governance_indicators:
        if indicator in module_lower:
            return "governance_reachable"
        # Also check if reachable from governance entry points
        for ep in reachable_from:
            if indicator in ep.lower():
                return "governance_reachable"

    for indicator in gate_indicators:
        if indicator in module_lower:
            return "gate_reachable"
        # Also check if reachable from gate entry points
        for ep in reachable_from:
            if indicator in ep.lower():
                return "gate_reachable"

    # Default to governance reachable if reachable from somewhere
    return "governance_reachable"

def analyze_modules():
    """Main analysis function."""
    src_path = "crates/franken-engine/src"

    # Step 1: Collect all modules
    modules = collect_modules(src_path)
    print(f"Found {len(modules)} modules")

    # Step 2: Parse dependencies for each module
    module_deps = {}
    for module_name, file_path in modules.items():
        deps = parse_module_dependencies(file_path)
        module_deps[module_name] = deps

    # Step 3: Define entry points
    entry_points = get_entry_points(modules)
    print(f"Entry points: {sorted(entry_points)}")

    # Step 4: Compute reachability from each entry point
    reachability = {}
    for entry_point in entry_points:
        reachable = compute_reachable_modules(entry_point, module_deps)
        reachability[entry_point] = reachable

    # Step 5: Classify each module
    analysis = {
        "modules": {},
        "entry_points": sorted(entry_points),
        "summary": {
            "total_modules": len(modules),
            "core_reachable": 0,
            "governance_reachable": 0,
            "gate_reachable": 0,
            "island_modules": 0
        }
    }

    for module_name, file_path in modules.items():
        reachable_from = set()

        for entry_point, reachable_modules in reachability.items():
            if module_name in reachable_modules:
                reachable_from.add(entry_point)

        classification = classify_module(module_name, reachable_from)

        module_info = {
            "name": module_name,
            "file_path": file_path,
            "dependencies": sorted(module_deps.get(module_name, set())),
            "classification": classification,
            "reachable_from": sorted(reachable_from)
        }

        analysis["modules"][module_name] = module_info

        # Update summary
        analysis["summary"][classification] += 1

    # Step 6: Output results
    output_file = "docs/architecture/module_classification.json"
    with open(output_file, 'w') as f:
        json.dump(analysis, f, indent=2, sort_keys=True)

    print("\nModule Classification Summary:")
    summary = analysis["summary"]
    print(f"Total modules: {summary['total_modules']}")
    print(f"Core reachable: {summary['core_reachable']}")
    print(f"Governance reachable: {summary['governance_reachable']}")
    print(f"Gate reachable: {summary['gate_reachable']}")
    print(f"Island modules: {summary['island_modules']}")

    # Show examples of each category
    print("\nExamples by category:")
    categories = {
        "Core reachable": "core_reachable",
        "Governance reachable": "governance_reachable",
        "Gate reachable": "gate_reachable",
        "Island": "island"
    }

    for category_name, category_key in categories.items():
        examples = [
            name for name, info in analysis["modules"].items()
            if info["classification"] == category_key
        ][:5]
        print(f"{category_name}: {examples}")

    print(f"\nResults written to {output_file}")
    return analysis

if __name__ == "__main__":
    analyze_modules()

File: scripts/analysis/analyze_module_deps.rs
#!/usr/bin/env rust-script
//! Module dependency analyzer for franken-engine
//!
//! This script analyzes the dependency graph of all modules in the franken-engine crate
//! and classifies them based on reachability from entry points.

use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};
use regex::Regex;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum ModuleClassification {
    CoreReachable,     // Reachable from execution_orchestrator.rs and core runtime
    GovernanceReachable, // Reachable from governance/control plane modules
    GateReachable,     // Reachable from gate/verification modules
    Island,            // Not reachable from any entry point
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ModuleInfo {
    pub name: String,
    pub file_path: String,
    pub dependencies: BTreeSet<String>,
    pub classification: ModuleClassification,
    pub reachable_from: BTreeSet<String>, // Entry points that can reach this module
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ModuleAnalysis {
    pub modules: BTreeMap<String, ModuleInfo>,
    pub entry_points: BTreeSet<String>,
    pub summary: ModuleSummary,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ModuleSummary {
    pub total_modules: usize,
    pub core_reachable: usize,
    pub governance_reachable: usize,
    pub gate_reachable: usize,
    pub island_modules: usize,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let src_path = Path::new("crates/franken-engine/src");

    // Step 1: Collect all modules
    let modules = collect_modules(src_path)?;
    println!("Found {} modules", modules.len());

    // Step 2: Parse dependencies for each module
    let mut module_deps: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
    for (module_name, file_path) in &modules {
        let deps = parse_module_dependencies(file_path)?;
        module_deps.insert(module_name.clone(), deps);
    }

    // Step 3: Define entry points
    let entry_points = get_entry_points(&modules);
    println!("Entry points: {:?}", entry_points);

    // Step 4: Build dependency graph and classify modules
    let mut analysis = ModuleAnalysis {
        modules: BTreeMap::new(),
        entry_points: entry_points.clone(),
        summary: ModuleSummary {
            total_modules: modules.len(),
            core_reachable: 0,
            governance_reachable: 0,
            gate_reachable: 0,
            island_modules: 0,
        },
    };

    // Step 5: Compute reachability from each entry point
    let mut reachability: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
    for entry_point in &entry_points {
        let reachable = compute_reachable_modules(entry_point, &module_deps);
        reachability.insert(entry_point.clone(), reachable);
    }

    // Step 6: Classify each module
    for (module_name, file_path) in &modules {
        let mut reachable_from = BTreeSet::new();

        for (entry_point, reachable_modules) in &reachability {
            if reachable_modules.contains(module_name) {
                reachable_from.insert(entry_point.clone());
            }
        }

        let classification = classify_module(module_name, &reachable_from);

        let module_info = ModuleInfo {
            name: module_name.clone(),
            file_path: file_path.display().to_string(),
            dependencies: module_deps.get(module_name).cloned().unwrap_or_default(),
            classification: classification.clone(),
            reachable_from,
        };

        // Update summary
        match classification {
            ModuleClassification::CoreReachable => analysis.summary.core_reachable += 1,
            ModuleClassification::GovernanceReachable => analysis.summary.governance_reachable += 1,
            ModuleClassification::GateReachable => analysis.summary.gate_reachable += 1,
            ModuleClassification::Island => analysis.summary.island_modules += 1,
        }

        analysis.modules.insert(module_name.clone(), module_info);
    }

    // Step 7: Output results
    let output_file = "docs/architecture/module_classification.json";
    let json = serde_json::to_string_pretty(&analysis)?;
    fs::write(output_file, json)?;

    println!("\nModule Classification Summary:");
    println!("Total modules: {}", analysis.summary.total_modules);
    println!("Core reachable: {}", analysis.summary.core_reachable);
    println!("Governance reachable: {}", analysis.summary.governance_reachable);
    println!("Gate reachable: {}", analysis.summary.gate_reachable);
    println!("Island modules: {}", analysis.summary.island_modules);

    // Show some examples of each category
    println!("\nExamples by category:");
    for (category, modules) in [
        ("Core reachable", ModuleClassification::CoreReachable),
        ("Governance reachable", ModuleClassification::GovernanceReachable),
        ("Gate reachable", ModuleClassification::GateReachable),
        ("Island", ModuleClassification::Island),
    ] {
        let examples: Vec<_> = analysis.modules.values()
            .filter(|m| m.classification == modules)
            .take(5)
            .map(|m| &m.name)
            .collect();
        println!("{}: {:?}", category, examples);
    }

    println!("\nResults written to {}", output_file);

    Ok(())
}

fn collect_modules(src_path: &Path) -> Result<BTreeMap<String, PathBuf>, std::io::Error> {
    let mut modules = BTreeMap::new();

    for entry in fs::read_dir(src_path)? {
        let entry = entry?;
        let path = entry.path();

        if path.is_file() && path.extension().map_or(false, |ext| ext == "rs") {
            if let Some(stem) = path.file_stem() {
                if let Some(module_name) = stem.to_str() {
                    modules.insert(module_name.to_string(), path);
                }
            }
        }
    }

    // Also check bin directory for binary targets
    let bin_path = src_path.join("bin");
    if bin_path.exists() {
        for entry in fs::read_dir(bin_path)? {
            let entry = entry?;
            let path = entry.path();

            if path.is_file() && path.extension().map_or(false, |ext| ext == "rs") {
                if let Some(stem) = path.file_stem() {
                    if let Some(module_name) = stem.to_str() {
                        modules.insert(format!("bin_{}", module_name), path);
                    }
                }
            }
        }
    }

    Ok(modules)
}

fn parse_module_dependencies(file_path: &Path) -> Result<BTreeSet<String>, std::io::Error> {
    let content = fs::read_to_string(file_path)?;
    let mut deps = BTreeSet::new();

    // Match use statements that reference crate modules
    let use_regex = Regex::new(r"use\s+(?:crate::)?([a-zA-Z_][a-zA-Z0-9_]*)")
        .expect("Invalid regex");

    // Match mod statements
    let mod_regex = Regex::new(r"mod\s+([a-zA-Z_][a-zA-Z0-9_]*)")
        .expect("Invalid regex");

    for cap in use_regex.captures_iter(&content) {
        if let Some(module) = cap.get(1) {
            deps.insert(module.as_str().to_string());
        }
    }

    for cap in mod_regex.captures_iter(&content) {
        if let Some(module) = cap.get(1) {
            deps.insert(module.as_str().to_string());
        }
    }

    Ok(deps)
}

fn get_entry_points(modules: &BTreeMap<String, PathBuf>) -> BTreeSet<String> {
    let mut entry_points = BTreeSet::new();

    // Primary entry points
    entry_points.insert("execution_orchestrator".to_string());
    entry_points.insert("lib".to_string()); // lib.rs public API

    // Binary targets are also entry points
    for module_name in modules.keys() {
        if module_name.starts_with("bin_") {
            entry_points.insert(module_name.clone());
        }
    }

    entry_points
}

fn compute_reachable_modules(
    entry_point: &str,
    module_deps: &BTreeMap<String, BTreeSet<String>>
) -> BTreeSet<String> {
    let mut visited = BTreeSet::new();
    let mut to_visit = vec![entry_point.to_string()];

    while let Some(current) = to_visit.pop() {
        if visited.contains(&current) {
            continue;
        }

        visited.insert(current.clone());

        if let Some(deps) = module_deps.get(&current) {
            for dep in deps {
                if !visited.contains(dep) {
                    to_visit.push(dep.clone());
                }
            }
        }
    }

    visited
}

fn classify_module(module_name: &str, reachable_from: &BTreeSet<String>) -> ModuleClassification {
    if reachable_from.is_empty() {
        return ModuleClassification::Island;
    }

    // Core reachable: reachable from execution_orchestrator or lib
    if reachable_from.contains("execution_orchestrator") || reachable_from.contains("lib") {
        return ModuleClassification::CoreReachable;
    }

    // Check if module or its entry points suggest governance/control plane
    let governance_indicators = [
        "control_plane", "governance", "policy", "scorecard", "audit",
        "compliance", "security", "attestation", "witness"
    ];

    let gate_indicators = [
        "gate", "verification", "conformance", "test", "benchmark",
        "harness", "validator", "checker"
    ];

    let module_lower = module_name.to_lowercase();

    for indicator in &governance_indicators {
        if module_lower.contains(indicator) ||
           reachable_from.iter().any(|ep| ep.to_lowercase().contains(indicator)) {
            return ModuleClassification::GovernanceReachable;
        }
    }

    for indicator in &gate_indicators {
        if module_lower.contains(indicator) ||
           reachable_from.iter().any(|ep| ep.to_lowercase().contains(indicator)) {
            return ModuleClassification::GateReachable;
        }
    }

    // Default to governance reachable if reachable from somewhere
    ModuleClassification::GovernanceReachable
}
